src/gwc/src/main/java/org/geoserver/gwc/dispatch/GwcServiceDispatcherCallback.java

```diff
@@ -29,21 +36,95 @@
         String context = request.getContext();
-        if (context == null || !context.startsWith("gwc/service")) {
+        if (context == null || !context.toLowerCase().contains("gwc/service")) {
             return null;
         }
 
-        Map<String, String> kvp = new HashMap<String, String>();
+        Map<String, String> kvp = new HashMap<>();
         kvp.put("service", "gwc");
         kvp.put("version", "1.0.0");
         kvp.put("request", "dispatch");
+
+        // if we are in the presence of virtual service we need to adapt the request
+        WorkspaceInfo localWorkspace = LocalWorkspace.get();
+        if (localWorkspace != null) {
+            // this is a virtual service request
+            String layerName = (String) request.getKvp().get("LAYER");
+            if (layerName == null) {
+                layerName = (String) request.getKvp().get("layer");
+            }
+            if (layerName != null) {
+                // we have a layer name as parameter we eed to adapt it (gwc doesn't care about workspaces)
+                layerName = CatalogConfiguration.removeWorkspacePrefix(layerName, catalog);
+                layerName = localWorkspace.getName() + ":" + layerName;
+                // we set the layer parameter with GWC expected name
+                kvp.put("LAYER", layerName);
+            }
+            // we need to setup a proper context path (gwc doesn't expect the workspace to be part of the URL)
+            request.setHttpRequest(new VirtualServiceRequest(request.getHttpRequest(), localWorkspace.getName(), layerName));
+        }
+
         request.setKvp(kvp);
         request.setRawKvp(kvp);
 
         return request;
     }
 
+    /**
+     * Helper wrapper that allow to match GWC expectations. GWC doesn't have the concept of workspaces,
+     * so he always expect a layer name to be prefixed by is workspace.
+     */
+    private final class VirtualServiceRequest extends HttpServletRequestWrapper {
+
+        private final String localWorkspaceName;
+        private final String layerName;
+
+        private final Map<String, String[]> parameters;
+
+        public VirtualServiceRequest(HttpServletRequest request, String localWorkspaceName, String layerName) {
+            super(request);
+            this.localWorkspaceName = localWorkspaceName;
+            this.layerName = layerName;
+            parameters = new HashMap<>(request.getParameterMap());
+            if (layerName != null) {
+                parameters.put("layer", new String[]{layerName});
+            }
+        }
+
+        @Override
+        public String getContextPath() {
+            // to GWC the workspace is part of the request context
+            return super.getContextPath() + "/" + localWorkspaceName;
+        }
+
+        @Override
+        public String getParameter(String name) {
+            if (layerName != null && name.equalsIgnoreCase("layer")) {
+                return layerName;
+            }
```

Comment:
Good question ! Let's wait for feedback on this http://osgeo-org.1560.x6.nabble.com/Styles-and-virtual-services-td5273216.html